chain/stmgr/execute.go

```diff
@@ -128,10 +128,17 @@ func (sm *StateManager) ExecutionTraceWithMonitor(ctx context.Context, ts *types
 }
 
 func (sm *StateManager) ExecutionTrace(ctx context.Context, ts *types.TipSet) (cid.Cid, []*api.InvocResult, error) {
+	if entry, ok := sm.execTraceCache.Get(ts); ok {
+		return entry.cid, entry.invocTrace, nil
```

Comment:
For better safety, you will (unfortunately) want to copy the invocation result objects. If you hand out a slice of pointers, the caller can mutate your objects.

Reply:
Ah, ok so on cache hit we need to do a deep copy of the cached `invocTrace` and return the copy

Reply:
Also, we also need to make a deep copy when we add to the cache since the caller can mutate that object as well. I assume we also have to deal with thread safety here as well so we need to manage a lock to make sure we have consistent view while making the deep copy.